AutoCollection/HeartBeat.ts

```diff
@@ -61,16 +70,79 @@ class HeartBeat {
             properties["appSrv_wsHost"] = process.env.WEBSITE_HOSTNAME || "";
         } else if (process.env.FUNCTIONS_WORKER_RUNTIME) { // Function apps
             properties["azfunction_appId"] = process.env.WEBSITE_HOSTNAME;
-        } 
-        this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
-        callback();
+        } else if (config && this._isVM === undefined) {
+            waiting = true;
+            this._getAzureComputeMetadata(config, () => {
+                if (this._isVM && Object.keys(this._vmData).length > 0) { // VM
+                    properties["azInst_vmId"] = this._vmData["vmId"] || "";
+                    properties["azInst_subscriptionId"] = this._vmData["subscriptionId"] || "";
+                    properties["azInst_osType"] = this._vmData["osType"] || "";
+                }
+                this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
+                callback();
+            });
+        }
+        if (!waiting) {
+            this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
+            callback();
+        }
     }
 
     public dispose() {
         HeartBeat.INSTANCE = null;
         this.enable(false);
         this._isInitialized = false;
     }
+
+    private _getAzureComputeMetadata(config: Config, callback: () => void) {
+        const metadataRequestUrl = `${AIMS_URI}?${AIMS_API_VERSION}&${AIMS_FORMAT}`;
+        const requestOptions = {
+            method: 'GET',
+            [AutoCollectHttpDependencies.disableCollectionRequestOption]: true,
+            headers: {
+                "Metadata": "true",
+            }
+        };
+
+        const req = Util.makeRequest(config, metadataRequestUrl, requestOptions, (res) => {
+            if (res.statusCode === 200) {
+                // Success; VM
+                this._isVM = true;
+                let virtualMachineData = "";
+                res.on('data', (data: any) => {
+                    virtualMachineData += data;
+                });
+                res.on('end', () => {
+                    this._vmData = this._isJSON(virtualMachineData) ? JSON.parse(virtualMachineData) : {};
```

Comment:
For the else case here, is it still guaranteed it is a VM? Maybe we can fill in some fallback defaults here for trying to figure out the OS, that we know it is a VM, etc

Reply:
I thought about how to reply your comment on this question, I added this check was in order to pass test... 😅 because the E2E test used a string "data" as default response data content, and failed JSON.parse, so added this to make sure the test won't break. But I get what you mean, and read the doc but don't find any case that it won't return a JSON. Check this [sample](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service#sample-3-getting-more-information-about-the-vm-during-support-case). The `.../metadata/instance/compute?...` API is to get more information of the VM instance, and it returns a JSON object. So in here, if the status code is 200, it should return a JSON string... so I don't think we need a fallback? WDYT?

Reply:
In the Python [implementation](https://github.com/census-instrumentation/opencensus-python/blob/master/contrib/opencensus-ext-azure/opencensus/ext/azure/metrics_exporter/heartbeat_metrics/heartbeat.py#L105) there are certain exceptions that are thrown when the app is not a vm (the endpoint simply doesn't exist). This should be the same case for Node correct? I'm not sure how the http client you are using handles exceptions, but I'm curious if a status code is even returned?

Reply:
The status code should get returned, check this [sample ](https://github.com/microsoft/ApplicationInsights-node.js/blob/53b440bdcb2eaa1cd6a04f125533fbc5c2f0ab10/Library/CorrelationIdManager.ts#L54) of imds api for node js. So if the status code is not 200, the request fails that it's not a vm. So here in my implementation, if the request fails it never gets to line 116 to check if the response data is JSON or not. And I learned the if elseif else from code [here](https://github.com/microsoft/ApplicationInsights-node.js/blob/53b440bdcb2eaa1cd6a04f125533fbc5c2f0ab10/Library/CorrelationIdManager.ts#L54) from node sdk repo.

Reply:
I'm not referring to how other apps handle error cases with status code. I'm asking about the functionality of the `http` library that is used in NodeJS (which `Util.makeRequests calls under the hood). Looking at the [docs](https://nodejs.org/api/http.html#http_http_request_options_callback) it looks like all errors are piped to `req.on('error...` so it should be handled already.